core/src/test/scala/kafka/raft/KafkaMetadataLogTest.scala

```diff
@@ -0,0 +1,333 @@
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.raft
+
+import java.io.File
+import java.nio.file.Files
+import java.nio.file.Path
+import kafka.log.Log
+import kafka.log.LogManager
+import kafka.log.LogTest
+import kafka.server.BrokerTopicStats
+import kafka.server.LogDirFailureChannel
+import kafka.utils.MockTime
+import kafka.utils.TestUtils
+import org.apache.kafka.common.TopicPartition
+import org.apache.kafka.common.errors.OffsetOutOfRangeException
+import org.apache.kafka.common.record.CompressionType
+import org.apache.kafka.common.record.MemoryRecords
+import org.apache.kafka.common.record.SimpleRecord
+import org.apache.kafka.common.utils.Utils
+import org.apache.kafka.raft.LogAppendInfo
+import org.apache.kafka.raft.LogOffsetMetadata
+import org.apache.kafka.raft.OffsetAndEpoch
+import org.apache.kafka.raft.ReplicatedLog
+import org.apache.kafka.snapshot.Snapshots
+import org.junit.After
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertThrows
+import org.junit.Assert.assertTrue
+import org.junit.Before
+import org.junit.Test
+
+final class KafkaMetadataLogTest {
+  import KafkaMetadataLogTest._
+
+  var tempDir: File = null
+  val mockTime = new MockTime()
+
+  @Before
+  def setUp(): Unit = {
+    tempDir = TestUtils.tempDir()
+  }
+
+  @After
+  def tearDown(): Unit = {
+    Utils.delete(tempDir)
+  }
+
+  @Test
+  def testCreateSnapshot(): Unit = {
+    val topicPartition = new TopicPartition("cluster-metadata", 0)
+    val numberOfRecords = 10
+    val epoch = 0
+    val snapshotId = new OffsetAndEpoch(numberOfRecords, epoch)
+    val log = buildMetadataLog(tempDir, mockTime, topicPartition)
+
+    append(log, numberOfRecords, epoch)
+    log.updateHighWatermark(new LogOffsetMetadata(numberOfRecords))
+
+    TestUtils.resource(log.createSnapshot(snapshotId)) { snapshot =>
+      snapshot.freeze()
+    }
+
+    TestUtils.resource(log.readSnapshot(snapshotId).get()) { snapshot =>
+      assertEquals(0, snapshot.sizeInBytes())
+    }
+  }
+
+  @Test
+  def testReadMissingSnapshot(): Unit = {
+    val topicPartition = new TopicPartition("cluster-metadata", 0)
+    val log = buildMetadataLog(tempDir, mockTime, topicPartition)
+
+    assertFalse(log.readSnapshot(new OffsetAndEpoch(10, 0)).isPresent)
```

Comment:
nit: doesn't matter too much for deterministic tests like this, but I think this is a nicer pattern:
```scala
assertEquals(Optional.empty(), log.readSnapshot(new OffsetAndEpoch(10, 0)))
```
Then the assertion message tells you what the value was.